gpu-switch.sh

```diff
@@ -0,0 +1,311 @@
+#!/bin/bash
+# gpu-switch.sh - Complete GPU management script for Docker containers
+# Supports multiple GPUs, optional GPU usage, and handles audio devices
+
+# Set up logging with timestamps
+log() {
+    local timestamp=$(date '+%Y-%m-%d %H:%M:%S')
+    logger "GPU-SWITCH [$timestamp]: $1"
+    echo "GPU-SWITCH [$timestamp]: $1" >&2
+}
+
+error_exit() {
+    log "ERROR: $1"
+    exit 1
+}
+
+# Store GPU and Audio pairs
+declare -A GPU_PAIRS
+
+check_gpu_needed() {
+    if [ -z "$NVIDIA_VISIBLE_DEVICES" ]; then
+        log "No GPUs specified in NVIDIA_VISIBLE_DEVICES - skipping GPU management"
+        return 1
+    fi
+    # Skip if NVIDIA_VISIBLE_DEVICES is set to none, None, or NONE
+    if [[ "${NVIDIA_VISIBLE_DEVICES,,}" == "none" ]]; then
+        log "NVIDIA_VISIBLE_DEVICES set to none - skipping GPU management"
+        return 1
+    fi
+    return 0
+}
+
+# Convert any GPU identifier to PCI address
+convert_to_pci_address() {
```

Comment:
Incorporating these improvements, here’s the final optimized convert_to_pci_address fu


```

convert_to_pci_address() {
    local device="$1"
    local gpu_address=""

    if [[ "$device" =~ ^[0-9]+$ || "$device" =~ ^GPU-.*$ ]]; then
        # Convert GPU index or UUID to PCI address
        gpu_address=$(nvidia-smi --id="$device" --query-gpu=gpu_bus_id --format=csv,noheader 2>/dev/null | tr -d '[:space:]')
    else
        # Direct PCI address provided
        gpu_address="$device"
    fi

    # Check for valid output
    if [ -z "$gpu_address" ]; then
        error_exit "Failed to get PCI address for device: $device"
    fi

    # Standardize format
    echo "$gpu_address" | sed -e 's/0000://' -e 's/\./:/g'
}

```

Reply:
ill take a look at implementing this thanks for the ideas on this as well

Reply:
merged with main